dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/ClientCore.cs

```diff
@@ -1571,4 +1540,154 @@ private void LogUsage(CompletionsUsage usage)
+        }
+
+        // If both behaviors are specified, we can't handle that.
+        if (executionSettings.FunctionChoiceBehavior is not null && executionSettings.ToolCallBehavior is not null)
+        {
+            throw new ArgumentException($"{nameof(executionSettings.ToolCallBehavior)} and {nameof(executionSettings.FunctionChoiceBehavior)} cannot be used together.");
+        }
+
+        // Set the tool choice to none. If we end up wanting to use tools, we'll set it to the desired value.
+        chatOptions.ToolChoice = ChatCompletionsToolChoice.None;
+        chatOptions.Tools.Clear();
+
+        // Handling new tool behavior represented by `PromptExecutionSettings.FunctionChoiceBehavior` property.
+        if (executionSettings.FunctionChoiceBehavior is { } functionChoiceBehavior)
+        {
+            result = this.ConfigureFunctionCalling(requestIndex, kernel, chatOptions, functionChoiceBehavior);
+        }
+        // Handling old-style tool call behavior represented by `OpenAIPromptExecutionSettings.ToolCallBehavior` property.
+        else if (executionSettings.ToolCallBehavior is { } toolCallBehavior)
+        {
+            result = this.ConfigureFunctionCalling(requestIndex, kernel, chatOptions, toolCallBehavior);
+        }
+
+        // Having already sent tools and with tool call information in history, the service can become unhappy "Invalid 'tools': empty array. Expected an array with minimum length 1, but got an empty array instead."
+        // if we don't send any tools in subsequent requests, even if we say not to use any.
+        // Similarly, if we say not to use any tool (ToolChoice = ChatCompletionsToolChoice.None) and dont provide any for the first request,
+        // the service fails with "'tool_choice' is only allowed when 'tools' are specified."
+        if (chatOptions.ToolChoice == ChatCompletionsToolChoice.None && chatOptions.Tools.Count == 0)
+        {
+            chatOptions.Tools.Add(s_nonInvocableFunctionTool);
+        }
+
+        return result;
+    }
+
+    private (bool? AllowAnyRequestedKernelFunction, int? MaximumAutoInvokeAttempts)? ConfigureFunctionCalling(int requestIndex, Kernel? kernel, ChatCompletionsOptions chatOptions, FunctionChoiceBehavior functionChoiceBehavior)
+    {
+        // Regenerate the tool list as necessary and getting other call behavior properties. The invocation of the function(s) could have augmented
+        // what functions are available in the kernel.
+        var config = functionChoiceBehavior.GetConfiguration(new() { Kernel = kernel });
+
+        (bool? AllowAnyRequestedKernelFunction, int? MaximumAutoInvokeAttempts) result = new()
+        {
+            AllowAnyRequestedKernelFunction = config.AllowAnyRequestedKernelFunction,
+            MaximumAutoInvokeAttempts = config.Options.AutoInvoke ? MaximumAutoInvokeAttempts : 0,
+        };
+
+        if (config.Choice == FunctionChoice.Required && requestIndex >= MaximumUseAttempts)
```

Comment:
If defining the FunctionChoiceBehavior as `required`, the OpenAI model will continue to run tools until the max auto invoke attempts are hit. How are you handling this, where the function result was returned to the model and tool should not be called again? Is the idea that the max auto invoke attempts should be 1 for a tool choice configured as `required`?

Reply:
As I am testing the `required` tool choice, no matter how I configure the max auto invoke attempts, I only get back a tool call as part of the Function Result. Versus `auto` where the model responds with a tool call, we send the tool call response and the model responds with the final answer. I'm curious to learn about how you're testing the `required` tool choice. Thanks.

Reply:
> If defining the FunctionChoiceBehavior as required, the OpenAI model will continue to run tools until the max auto invoke attempts are hit. How are you handling this, where the function result was returned to the model and tool should not be called again?

The 'if' statement on line 1606 ensures that the method exits before executing the code that advertises the function to the model, thanks to the 'return' statement. As a result, the required choice behavior advertises the function only once during the first request and stops doing so for subsequent requests.

> Is the idea that the max auto invoke attempts should be 1 for a tool choice configured as required?

In addition to the MaximumAutoInvokeAttempts, which caps the number of auto-invocations per single user interaction, there's also the MaximumUseAttempts constant, which specifies the number of requests per single user interaction that should include functions in the request.

> As I am testing the required tool choice, no matter how I configure the max auto invoke attempts, I only get back a tool call as part of the Function Result. Versus auto where the model responds with a tool call, we send the tool call response and the model responds with the final answer. I'm curious to learn about how you're testing the required tool choice. Thanks.

Add/advertise/include functions only to the first request to the model and stop doing so for all subsequent ones for the required function choice behavior.

Reply:
Thanks!